cpp/src/parquet/encoding.h

```diff
@@ -158,6 +158,10 @@ class Encoder {
   virtual Encoding::type encoding() const = 0;
 
   virtual void Put(const ::arrow::Array& values) = 0;
+  // Report the number of bytes before encoding that have been written
+  // to the encoder since the last report. Note that this call is not
+  // idempotent because it resets the internal counter.
+  virtual int64_t ReportUnencodedDataBytes() = 0;
```

Comment:
it might also be worth noting this isn't supported for bool type and may throw

Reply:
I think it is trivial to support this and dislike adding a type-specific function to a generic class.

Reply:
Sure, but it also doesn't make sense to maintain (and test) some code that has no use case, right?

At worse this can be put on a derived class :)

Reply:
I tried to add `std::enable_if_t<std::is_same_v<DType, ByteArrayType>, int64_t> ReportUnencodedDataBytes()` to TypedEncoder. However, enable_if does not support virtual function and it is still ugly if this is not a virtual function. So I decided to keep it at the base class and throw for non-ByteArray type. WDYT?

Reply:
Sounds good to me, yes.